document/js-api/index.bs

```diff
@@ -0,0 +1,840 @@
+    1. Let |imports| be an empty [=list=].
+    1. For each (|moduleBytes|, |nameBytes|, |type|) in [=module_imports=](|module|),
+        1. Let |moduleName| be |moduleBytes| [=UTF-8 decode without BOM or fail|decoded as UTF-8=].
+        1. Let |name| be |nameBytes| [=UTF-8 decode without BOM or fail|decoded as UTF-8=].
+        1. Assert: The previous operations are not failure, as |module| is [=valid=].
+        1. Let |kind| be the [=string value of the extern type=] |type|.
+        1. Let |obj| be a new {{ModuleImportDescriptor}} dictionary with {{ModuleImportDescriptor/module}} |moduleName|, {{ModuleImportDescriptor/name}} |name| and {{ModuleImportDescriptor/kind}} |kind|.
+        1. [=Append=] |obj| to the end of |imports|.
+    1. Return |imports|.
+</div>
+
+<div algorithm>
+    The <dfn method for="Module">customSections(moduleObject, sectionName)</dfn> method, when invoked, performs the following steps:
+    1. Let |bytes| be |moduleObject|.\[[Bytes]].
+    1. Let |customSections| be an empty [=list=] of {{ArrayBuffer}}s.
+    1. For each [=custom section=] |customSection| in the binary format of |bytes|,
+        1. Let |name| be the <code>name</code> of the custom section, [=UTF-8 decode without BOM or fail|decoded as UTF-8=].
+        1. Assert: |name| is not failure (|moduleObject|.\[[Module]] is [=valid=]).
+        1. If |name| equals |secondName| as string values,
+            1. [=Append=] a new {{ArrayBuffer}} containing a [=get a copy of the buffer source|copy of the bytes held by the buffer=] |bytes| for the range matched by this [=customsec=] production.
+    1. Return |customSections|.
+</div>
+
+<div algorithm>
+    The <dfn constructor for="Module">Module(bytes)</dfn> constructor, when invoked, performs the follwing steps:
+
+    1. Let |stableBytes| be a [=get a copy of the buffer source|copy of the bytes held by the buffer=] |bytes|.
+    1. [=Compile a WebAssembly module|Compile the WebAssembly module=] |stableBytes| and store the result as |module|.
+    1. If |module| is [=error=], throw a {{CompileError}} exception.
+    1. Return |module|.
+</div>
+
+<h3 id="instances">Instances</h3>
+
+<pre class="idl">
+[LegacyNamespace=WebAssembly, Constructor(Module module, optional object importObject), Exposed=(Window,Worker,Worklet)]
+interface Instance {
+  readonly attribute object exports;
+};
+</pre>
+
+<div algorithm>
+    The <dfn constructor for="Instance">Instance(module, importObject)</dfn> constructor, when invoked, [=instantiate a WebAssembly module|instantiates the WebAssembly module=] |module| importing |importObject| and returns the result.
+</div>
+
+<div algorithm>
+    The getter of the <dfn attribute for="Instance">exports</dfn> attribute of {{Instance}} returns the receiver's \[[Exports]] internal slot.
+</div>
+
+<h3 id="memories">Memories</h3>
+
+<pre class="idl">
+dictionary MemoryDescriptor {
+  required [EnforceRange] unsigned long initial;
+  [EnforceRange] unsigned long maximum;
```

Comment:
One other visible change is that [EnforceRange] throws `TypeError`s, not `RangeError`s which requires changing some of our [tests](https://github.com/WebAssembly/spec/blob/master/test/js-api/jsapi.js#L420).  It might be good to bring this up explicitly at the next CG meeting.

Reply:
Thanks for the attention to detail here. Do you think we're OK making this normative change, or do you think it has too much web compatibility risk? An alternative would be to make `initial` optional. I could imagine someone doing `new WebAssembly.Memory({})` to create an empty memory. There might be some other cases that change, but this one seems particularly plausible.

Reply:
In [bug 1418195](https://bugzilla.mozilla.org/show_bug.cgi?id=1418195) I landed a patch to conservatively throw on NaN (to see if anything would break).  So far I haven't heard a peep.  But it's good to hear that, if this was a problem, we could simply make it `optional` and keep using [EnforceRange].  Assuming I don't hear any bug reports, it makes sense to go with the stricter thing for now.  Might be good to keep a list of known normative changes (including these exception types) so the group can discuss.

Reply:
That change in SpiderMonkey looks is pretty recent; I still feel inclined to remove `required` here as the change isn't really strongly motivated in the first place.

Reply:
Specifically, I'm thinking of a compromise where we remove "required" in the dictionary but not make the other arguments optional and still throw in those cases (as well as throwing on NaN). What would you think of that?

Reply:
I understand what "remove required" means in WebIDL terms, but what does the rest of the sentence mean?

Reply:
Oh, interestingly there's an IDL restriction here: Dictionaries with no required members must be optional arguments. This would mean that `new WebAssembly.Memory()` would work as well with my suggestion, whereas it doesn't currently work in Firefox.

Reply:
Backing out [bug 1418195](https://bugzilla.mozilla.org/show_bug.cgi?id=1418195) is no big deal; it's still Nightly-only until mid-January.  So with that change, is the only semantic change throwing on NaN?

Reply:
Actually, given the IDL restriction, if you're comfortable leaving it in for a little bit longer, this could give us some useful data.